File: 64. Minimum Path Sum.py
# 64. Minimum Path Sum.

# Topic: Array, Dynamic Programming, Matrix.

'''
# Task:
-----------
Given a m x n grid filled with non-negative numbers, find a path from top left to bottom right, which 
minimizes the sum of all numbers along its path.

Note: You can only move either down or right at any point in time.

Example 1:
Input: grid = [[1,3,1],[1,5,1],[4,2,1]]
Output: 7
Explanation: Because the path 1 → 3 → 1 → 1 → 1 minimizes the sum.

Example 2:
Input: grid = [[1,2,3],[4,5,6]]
Output: 12

Constraints:
m == grid.length
n == grid[i].length
1 <= m, n <= 200
0 <= grid[i][j] <= 200

# Testcase:
------------------
[[1,3,1],[1,5,1],[4,2,1]]
[[1,2,3],[4,5,6]]

# Code:
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
      
'''
# Solution:
class Solution:
    def minPathSum(self, grid: list[list[int]]) -> int:
        if not grid:
            return 0

        m, n = len(grid), len(grid[0])
        dp = [[0] * n for _ in range(m)]

        for i in range(m):
            for j in range(n):
                if i == 0 and j == 0:
                    dp[i][j] = grid[i][j]  # Starting point
                elif i == 0:
                    dp[i][j] = dp[i][j-1] + grid[i][j]  # Top row
                elif j == 0:
                    dp[i][j] = dp[i-1][j] + grid[i][j]  # Leftmost column
                else:
                    dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j]  # Other cells

        return dp[m-1][n-1]

# Testcases
sol = Solution()
print(sol.minPathSum([[1,3,1],[1,5,1],[4,2,1]]))  # Output: 7
print(sol.minPathSum([[1,2,3],[4,5,6]]))  # Output: 12


# Description 
'''
To solve this problem, we will use Dynamic Programming (DP), which is a common technique for problems 
like this where we need to find an optimal solution based on decisions made at each step.

We will create a DP table of the same size as the grid and fill it with the minimum sum needed to reach 
each cell. The minimum sum to reach a cell is the value of the cell plus the minimum of the sums needed
to reach the cells above and to the left of it. The top-left cell is the starting point, so its minimum 
sum is its own value. For the cells in the top row and leftmost column, we only add the value of the 
previous cell in the respective row or column since there's only one way to reach them.

In this code:
- We first check if the grid is empty.
- m and n are the dimensions of the grid.
- dp is the DP table initialized with zeroes.
- We iterate over each cell in the grid, filling in the dp table with the minimum sum to reach that cell.
- The final answer is the value in the bottom-right cell of the dp table, which represents the minimum 
path sum to reach that point from the top-left corner.

'''

File: Cpp/64. Minimum Path Sum.cpp
// 64. Minimum Path Sum.

// Topic:Array,Dynamic Programming,Matrix.

/*
### Task:
Given a m x n grid filled with non-negative numbers, find a path from top left to bottom right, which minimizes the sum 
of all numbers along its path.

Note: You can only move either down or right at any point in time.

Example 1:
Input: grid = [[1,3,1],[1,5,1],[4,2,1]]
Output: 7
Explanation: Because the path 1 → 3 → 1 → 1 → 1 minimizes the sum.

Example 2:
Input: grid = [[1,2,3],[4,5,6]]
Output: 12

Constraints:
m == grid.length
n == grid[i].length
1 <= m, n <= 200
0 <= grid[i][j] <= 200


### Testcase:
[[1,3,1],[1,5,1],[4,2,1]]
[[1,2,3],[4,5,6]]

### Code:
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        
    }
};

*/
// Solution: -----------------------------------------------

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector<vector<int>> dp(m, vector<int>(n, grid[0][0]));

        // Initialize the first row and first column of the DP table
        for (int i = 1; i < m; i++) {
            dp[i][0] = dp[i - 1][0] + grid[i][0];
        }
        for (int j = 1; j < n; j++) {
            dp[0][j] = dp[0][j - 1] + grid[0][j];
        }

        // Compute the minimum path sum for each cell
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                // For each cell, choose the minimum path sum from the top or left cell and add the current cell's value
                dp[i][j] = min(dp[i - 1][j], dp[i][j - 1]) + grid[i][j];
            }
        }

        // The bottom-right cell contains the minimum path sum
        return dp[m - 1][n - 1];
    }
};

// Description: ============================================
/*
To solve the "Minimum Path Sum" problem, we can use dynamic programming. The idea is to update each cell in the grid with the minimum sum to reach that cell from the top left corner. We only need to consider two directions for each cell: from the left (i, j-1) and from the top (i-1, j). The minimum path sum for a cell (i, j) is the value in (i, j) plus the minimum of the sums of the cells from which we can arrive at (i, j), which are (i-1, j) and (i, j-1).

Here's a step-by-step approach to implement this:

1. Create a 2D DP array of the same size as the input grid to store the minimum sum path to reach each cell.
2. Initialize the first cell (top-left corner) of the DP array with the first cell of the grid since that's where we start.
3. Iterate through the grid row by row, and within each row column by column, updating the DP array:
   - For each cell, calculate the minimum path sum to reach that cell by considering the minimum path sum to reach the cell directly above and to the left, then add the current cell's value.
   - For the first row and first column, we only have one choice: move right from the left cell or move down from the top cell, respectively.
4. The bottom-right corner of the DP array will contain the minimum path sum from the top-left corner to the bottom-right corner.

### Explanation:
- **Line 3-4:** We get the dimensions of the grid (`m` and `n`) and initialize the DP table with the value of the first cell of the grid.
- **Lines 7-10:** We fill in the first column of the DP table. Each cell in the first column can only be reached from the cell directly above, so we add the current cell's value to the sum of the path to the cell above.
- **Lines 11-14:** Similarly, we fill in the first row of the DP table. Each cell in the first row can only be reached from the cell directly to the left, so we add the current cell's value to the sum of the path to the cell to the left.
- **Lines 17-21:** We iterate over the rest of the grid, computing the minimum path sum for each cell as the minimum of the sums of the top and left cells plus the current cell's value.
- **Line 24:** The minimum path sum to reach the bottom-right corner of the grid is contained in the bottom-right cell of the DP table, which we return as the solution.

*/
